logic/LogicCache.java

```diff
@@ -31,23 +31,30 @@
 
 public class LogicCache {
 
-    private final CommonCache<GraphTraversal.Type, Optional<Map<Identifier.Variable.Retrievable, Set<Label>>>> typeInferenceCache;
     private final CommonCache<String, Rule> ruleCache;
     private final CommonCache<Concludable, Map<Rule, Set<Unifier>>> unifiers;
+    private final CommonCache<GraphTraversal.Type, Optional<Map<Identifier.Variable.Retrievable, Set<Label>>>> typeInferenceCache;
+    private final CommonCache<GraphTraversal.Type, Boolean> queryCoherenceCache;
```

Comment:
we have two caches: the type inference cache, and query coherence cache. Query coherence is simply a boolean about whether a query is illegal or not. Type inference is the same as the previous cache